### PR TITLE
[ISSUE-867][BUG] Create writer failed should report non-critical error instead of critical error

### LINKED ISSUE #867: [BUG] Create Writer failed should not report critical error, should report non-critical error
### How to reproduce the bug?
Steps to reproduce the bug.

/cc @who-need-to-know

/assign @who-can-solve-this-bug

### PR DESCRIPTION
**Pull request by nafiyAix**
### What changes were proposed in this pull request?
Create writer failed should report non-critical error instead of critical error

### Why are the changes needed?
To avoid the problem that execute createWriter() method twice will make all Flusher‘s thread stop and hang on the normal execution of worker.

### What are the items that need reviewer attention?
Could we just ignore the request if we check that the target file of fileWriter already exists. And if there is other I/O issue when creating filewriter, we could handle it by previous way.

### Related issues.
close #867 

### How was this patch tested?

/cc @related-reviewer

/assign @main-reviewer

### CODE REVIEW

File: worker/src/main/scala/org/apache/celeborn/service/deploy/worker/storage/StorageManager.scala [hunk 1]
```diff
@@ -333,9 +334,20 @@ final private[worker] class StorageManager(conf: CelebornConf, workerSource: Abs
           logDebug(s"location $location set disk hint to ${location.getStorageInfo} ")
           return fileWriter
         } catch {
-          case t: Throwable =>
+          case fe: FileAlreadyExistsException =>
```
Comment (AngersZhuuuu) [inline worker/src/main/scala/org/apache/celeborn/service/deploy/worker/storage/StorageManager.scala:337]: If `FileAlreadyExistsException`, we don't need to report error, right?

File: worker/src/main/scala/org/apache/celeborn/service/deploy/worker/storage/StorageManager.scala [hunk 2]
```diff
@@ -333,9 +334,20 @@ final private[worker] class StorageManager(conf: CelebornConf, workerSource: Abs
...
+            logError(s"Target file ${file.getAbsolutePath} of fileWriter already exists.")
+            exception = new IOException(fe)
+            deviceMonitor.reportNonCriticalError(mountPoint, fe)
+          case rsse: RssException =>
```
Comment (AngersZhuuuu) [inline worker/src/main/scala/org/apache/celeborn/service/deploy/worker/storage/StorageManager.scala:341]: If we create files failed  just caused by 
```
1. Not enough file handler
2. Not enough space
```
We should not report the critical error to cause all fileWriter in this device fail.

File: worker/src/main/scala/org/apache/celeborn/service/deploy/worker/storage/StorageManager.scala [hunk 3]
```diff
@@ -333,12 +346,18 @@ final private[worker] class StorageManager(conf: CelebornConf, workerSource: Abs
           logDebug(s"location $location set disk hint to ${location.getStorageInfo} ")
           return fileWriter
         } catch {
+          case fe: FileAlreadyExistsException =>
+            logError("Failed to create fileWriter because of existed file", fe)
```
Comment (AngersZhuuuu) [inline worker/src/main/scala/org/apache/celeborn/service/deploy/worker/storage/StorageManager.scala:350]: miss throw exception?

File: worker/src/main/scala/org/apache/celeborn/service/deploy/worker/storage/StorageManager.scala [hunk 4]
```diff
@@ -333,12 +346,18 @@ final private[worker] class StorageManager(conf: CelebornConf, workerSource: Abs
...
-              s"Create FileWriter for ${file.getAbsolutePath} of mount $mountPoint failed, report to DeviceMonitor",
+              s"Create FileWriter for ${file.getAbsolutePath} of mount $mountPoint " +
+                s"failed, report to DeviceMonitor",
               t)
             exception = new IOException(t)
-            deviceMonitor.reportDeviceError(mountPoint, exception, DiskStatus.READ_OR_WRITE_FAILURE)
+            deviceMonitor.reportNonCriticalError(
+              mountPoint,
+              exception,
+              DiskStatus.READ_OR_WRITE_FAILURE)
```
Comment (AngersZhuuuu) [inline worker/src/main/scala/org/apache/celeborn/service/deploy/worker/storage/StorageManager.scala:360]: Add a new type .....

Comment (nafiyAix) [reply worker/src/main/scala/org/apache/celeborn/service/deploy/worker/storage/StorageManager.scala:360]: I think currently this type is sufficient to describe the possible issues, `ObservedDevice` and `Flusher` also use this type to present all unexpected IOException may meet , if we add a new type will it a bit redundant?

File: worker/src/main/scala/org/apache/celeborn/service/deploy/worker/storage/StorageManager.scala [hunk 5]
```diff
@@ -333,12 +346,19 @@ final private[worker] class StorageManager(conf: CelebornConf, workerSource: Abs
...
```
Comment (AngersZhuuuu) [inline worker/src/main/scala/org/apache/celeborn/service/deploy/worker/storage/StorageManager.scala:361]: We should add a new one flag?

File: worker/src/main/scala/org/apache/celeborn/service/deploy/worker/storage/StorageManager.scala [hunk 6]
```diff
@@ -147,13 +147,25 @@ final private[worker] class StorageManager(conf: CelebornConf, workerSource: Abs
...
     }
   }
 
+  override def notifyNonCriticalError(mountPoint: String, diskStatus: DiskStatus): Unit =
+    this.synchronized {
+      if (diskStatus == DiskStatus.READ_OR_WRITE_FAILURE) {
+        val dir = new File(mountPoint)
+        // TODO: don't destroy exist fileWriter, but will remain fileWriter cause memory leak or other issue?
+        // And do we need to also clean the diskOperators here?
+        workingDirWriters.remove(dir)
```
Comment (AngersZhuuuu) [inline worker/src/main/scala/org/apache/celeborn/service/deploy/worker/storage/StorageManager.scala:165]: When to add back?

**Comment by AngersZhuuuu** [comment]
截屏2022-10-31 下午2 08 05

File: worker/src/main/scala/org/apache/celeborn/service/deploy/worker/storage/DeviceMonitor.scala [hunk 7]
```diff
@@ -93,6 +95,13 @@ class LocalDeviceMonitor(
         })
       }
 
+    def notifyObserversOnNonCriticalError(mountPoint: String, diskStatus: DiskStatus): Unit =
+      this.synchronized {
+        diskInfos.get(mountPoint).setStatus(diskStatus)
+        val tmpObservers = new util.HashSet[DeviceObserver](observers)
+        tmpObservers.asScala.foreach(ob => ob.notifyNonCriticalError(mountPoint, diskStatus))
```
Comment (AngersZhuuuu) [inline worker/src/main/scala/org/apache/celeborn/service/deploy/worker/storage/DeviceMonitor.scala:102]:
```
foreach(_.notifyxxxx)

**Comment by AngersZhuuuu** [comment]
ping @FMX @waitinfuture

File: worker/src/main/scala/org/apache/celeborn/service/deploy/worker/storage/StorageManager.scala [hunk 8]
```diff
@@ -154,6 +154,17 @@ final private[worker] class StorageManager(conf: CelebornConf, workerSource: Abs
     }
   }
 
+  override def notifyNonCriticalError(mountPoint: String, diskStatus: DiskStatus): Unit =
```
Comment (AngersZhuuuu) [inline worker/src/main/scala/org/apache/celeborn/service/deploy/worker/storage/StorageManager.scala:157]: Remove  this method should be ok

**Review by AngersZhuuuu — APPROVED** [review]

**Comment by FMX** [comment]
\regression

**Review by waitinfuture — APPROVED** [review]
LGTM, thanks!

**Comment by waitinfuture** [comment]
> /regression

Regression passed
https://github.com/alibaba/celeborn/actions/runs/3361161115/jobs/5571216042
It would be much better if we can auto-generate regression url :) @FMX

**Pull request merged**